src/rust/lib/system/web/src/lib.rs

```diff
@@ -481,3 +483,79 @@ pub mod traits {
     pub use super::NodeInserter;
     pub use super::NodeRemover;
 }
+
+/// Sleeps for the specified amount of time.
+///
+/// This function might sleep for slightly longer than the specified duration but never less.
+///
+/// This function is an async version of std::thread::sleep.
+#[cfg(target_arch = "wasm32")]
+pub async fn sleep(duration:Duration) {
+    use wasm_bindgen_futures::JsFuture;
+
+    let performance = performance();
+    let call_time   = performance.now();
+    let future : JsFuture = js_sys::Promise::new(&mut |resolve:Function,_| {
+        let time_elapsed_from_call = ((performance.now() - call_time) * 1000.0) as i32;
+        let duration               = duration.as_millis() as i32;
+        let duration               = (duration - time_elapsed_from_call).max(0);
+        let window                 = window();
+        let err                    = "setTimeout failed";
+        window.set_timeout_with_callback_and_timeout_and_arguments_0(&resolve,duration).expect(err);
+    }).into();
+    future.await.expect("setTimeout future failed");
+}
+
+#[cfg(not(target_arch = "wasm32"))]
+pub use async_std::task::sleep;
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use wasm_bindgen_test::wasm_bindgen_test;
+    use wasm_bindgen_test::wasm_bindgen_test_configure;
+
+    wasm_bindgen_test_configure!(run_in_browser);
+
+    #[cfg(target_arch = "wasm32")]
+    mod helpers {
+        type Instant = f64;
+
+        pub fn now() -> Instant {
+            super::performance().now()
+        }
+
+        pub fn elapsed(instant: Instant) -> f64 {
+            super::performance().now() - instant
+        }
+    }
+
+    #[cfg(not(target_arch = "wasm32"))]
+    mod helpers {
+        use std::time::Instant;
+
+        pub fn now() -> Instant {
+            Instant::now()
+        }
+
+        pub fn elapsed(instant: Instant) -> f64 {
+            instant.elapsed().as_secs_f64()
+        }
+    }
+
+    #[wasm_bindgen_test(async)]
+    async fn async_sleep() {
+        let instant = helpers::now();
+        sleep(Duration::new(1,0)).await;
+        assert!(helpers::elapsed(instant) >= 1.0);
+        sleep(Duration::new(2,0)).await;
+        assert!(helpers::elapsed(instant) >= 3.0);
+    }
```

Comment:
I would also check if the time was not too big, like `>=1.0` but also `<= 1.0 + epsilon`

Reply:
@wdanilo I think it's problematic. `setTimeout` will always run after the specified duration, but the exact moment is not predictable. What if CI's load balancer makes `setTimeout` take much more time to be executed than expected? We can either leave the test as it is or, if you really want it, we can define an bigger threshold to test it, and not a small number (epsilon).

Reply:
I'm not sure about this, it wouldn't exactly test our code correctness but also the CI machine's load. Race conditions always scare me.
(if we wanted to be very-correct, we should mock the instant/set_timeout api, but it seems to be overkill at this stage)